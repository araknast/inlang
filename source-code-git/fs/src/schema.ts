export type Filesystem = {
	writeFile: (path: string, content: string) => Promise<void>
	readFile: (path: string) => Promise<string | null>
	readdir: (path: string) => Promise<string[] | null>
	mkdir: (path: string) => Promise<void>
	/**
	 * Serializes the filesystem to a JSON string.
	 *
	 * @example
	 *   const result = fs.toJson()
	 *   >> { "file1.txt": "content", "file2.txt": "content" }
	 */
	toJson: (args: {
		/**
		 * Takes a list of glob patterns to include.
		 *
		 * @example
		 *   fs.toJson({ include: ["**\/*.json", "**\/*.md"] })
		 */
		include?: string[]
		/**
		 * Takes a list of glob patterns to exclude.
		 *
		 * @example
		 *   fs.toJson({ exclude: ["**\/node_modules"] })
		 */
		exclude?: string[]
	}) => Promise<Record<string, any>>
}
