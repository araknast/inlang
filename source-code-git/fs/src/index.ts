export type { Filesystem } from "./schema.js"
export { MemoryFs } from "./memory/index.js"
