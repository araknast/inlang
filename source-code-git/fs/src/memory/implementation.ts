import type { Filesystem } from "../schema.js"

type FileData = string
type Directory = Map<string, Inode>
type Inode = FileData | Directory

export class MemoryFs implements Filesystem {
	root: Directory
	specialPaths: string[]

	constructor(root: Directory = new Map()) {
		this.root = root
		this.specialPaths = ["", ".", ".."]
		this.root.set("", this.root) 
		this.root.set(".", this.root) 
		this.root.set("..", this.root) 
	}

	_newDir(parentDir: Directory) {
		// A circular reference allows for simple handling of leading and
		// trailing slashes, as well as "." paths
		//	
		// Since these only exist internaly, there shouldn't be issues
		// with serialization
		const dir = new Map()
		dir.set("", dir)
		dir.set(".", dir)
		dir.set("..", parentDir)
		return dir
	}

	_followPath(path: string, makeParent: boolean = false): Inode | undefined {
		const pathList: string[] = path.split("/")
		let target: Inode | undefined = this.root
		let parentDir: Directory

		if (makeParent) {
			for (let path of pathList) {

				if (target instanceof Map) {
					parentDir = target
					target = target.get(path) ?? undefined
				} else break

				if (!target) {
					if (!makeParent) return undefined
					parentDir.set(path, this._newDir(parentDir))
					target = parentDir.get(path) ?? undefined
				}
			}
		} else {
			for (let path of pathList) {
				if (target instanceof Map) 
					target = target.get(path) ?? undefined
				else break

				if (!target) return undefined
			}
		}

		return target
	}

	dirname = (path: string): string  => path.split("/").slice(0, -1).join("/")
	basename = (path: string): string => path.split("/").slice(-1)[0]

	async writeFile(path: string, content: FileData) {
		const parentDir: Inode | undefined = this._followPath(this.dirname(path), true)
		if (parentDir instanceof Map) parentDir.set(this.basename(path), content)
	}

	async readFile(path: string): Promise<FileData | undefined> {
		const file: Inode | undefined = this._followPath(path)
		return typeof file === "string" ? file : undefined
	}

	async readdir(path: string): Promise<string[] | undefined> {
		const dir: Inode | undefined = this._followPath(path)
		if (dir instanceof Map)
			return [...dir.keys()].filter(x => !(this.specialPaths.includes(x)))
		else return
	}

	async mkdir(path: string) {
		this._followPath(path, true)
	}
	
	_dirToJson(dir: Directory): Record<string,any> {
		const json: Record<string, any> = {}
		for (let kv of dir) {
			if (this.specialPaths.includes(kv[0])) continue
			if (kv[1] instanceof Map) {
				json[kv[0]] = this._dirToJson(kv[1])
			} else if (typeof kv[1] === "string"){
				json[kv[0]] = kv[1]
			}
		}
		return json
	}

	async toJson(args: any): Promise<Record<string, any>> {
		return this._dirToJson(this.root)
	}

	static _dirFromJson(json: Record<string, any>): Directory {
		const dir: Directory = new Map()
		for (let kv of Object.entries(json)) {
			if (typeof kv[1] === "object") {
				dir.set(kv[0], MemoryFs._dirFromJson(kv[1]))
			} else if (typeof kv[1] === "string") {
				dir.set(kv[0], kv[1])
			}
		}
		return dir
	}

	static async fromJson(jsonString: string): Promise<Filesystem> {
		const json: Record<string, any> = JSON.parse(jsonString)
		return new MemoryFs(MemoryFs._dirFromJson(json))
	}
}
