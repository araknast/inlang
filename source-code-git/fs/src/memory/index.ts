export { memoryFs } from "./implementation.js"
