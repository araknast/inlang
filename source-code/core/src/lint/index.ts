export { lint } from "./linter.js"
export { getLintReports, hasLintReports } from "./query.js"
export { createLintRule, type LintRule, type LintRuleInitializer, type LintRuleId } from "./rule.js"
export { createLintRuleCollection, type RuleCollectionInitializer } from "./ruleCollection.js"
