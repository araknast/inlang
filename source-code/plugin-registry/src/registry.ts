import type { InlangPluginManifest } from "./schema.js"

export const plugins: InlangPluginManifest[] = [
	{
		id: "inlang-plugin-json",
		icon: "https://github.com/samuelstroschein/inlang-plugin-json/assets/icon.png",
		repository: "https://github.com/samuelstroschein/inlang-plugin-json",
		keywords: ["json", "i18next", "svelte-intl-precompile"],
	},
	{
		id: "plugin-i18next",
		icon: "https://github.com/samuelstroschein/inlang-plugin-json/assets/icon.png",
		repository: "https://github.com/samuelstroschein/inlang-plugin-json",
		keywords: ["json", "i18next", "svelte-intl-precompile"],
	},
	{
		id: "inlang-lint",
		icon: "https://github.com/samuelstroschein/inlang-plugin-json/assets/icon.png",
		repository: "https://github.com/samuelstroschein/inlang-plugin-json",
		keywords: ["json", "i18next", "svelte-intl-precompile"],
	},
]
