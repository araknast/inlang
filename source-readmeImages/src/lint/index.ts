// // todo
// const results = {}

import { http, raw } from "../api/index.js"
import { fs } from "memfs"
// // ---------------

// if (results['inlang/inlang'].time < h24ago){

// }
await raw.clone({
	fs: fs,
	http,
	dir: "/",
	url: `https://github.com/inlang/inlang`,
	singleBranch: true,
	depth: 1,
})

// const file = await fs.readFile("/inlang.config.js")
// const { defineConfig } = await import(file)
// const config = await defineConfig()
// results['inlang/inlang'] = {r: lint(config), time: Date.now()}

// const html = <Image {result}></Image>

// response.send(renderToImage(html))
async function cloneRepository(args: { fs: typeof import("memfs").fs }): Promise<Date | undefined> {
	// do shallow clone, get first commit and just one branch
	await raw.clone({
		fs: args.fs,
		http,
		dir: "/",
		url: `https://github.com/inlang/inlang`,
		singleBranch: true,
		depth: 1,
	})

	// triggering a side effect here to trigger a re-render
	// of components that depends on fs
	const date = new Date()
	return date
}
cloneRepository({ fs })
const file = await fs.readFile("/inlang.config.js", cloneRepository({ fs }))
console.log("123", file)
