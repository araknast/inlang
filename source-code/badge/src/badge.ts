import satori from "satori"
import { html } from "satori-html"
import clone from "./repo/clone.js"
import { Config, EnvironmentFunctions, initialize$import } from "@inlang/core/config"
import { getLintReports, lint } from "@inlang/core/lint"
import { Volume } from "memfs"
import type { fs as memfs } from "memfs"

// // const markup = (url: string) => html`<div style="color: black;">${url}</div>`
// const markup = html`<div
// 	style="display: flex; background-color: navy; height: 300px; width: 300px"
// ></div>`
// // See https://github.com/vercel/satori#documentation

// const font = readFileSync(new URL("./assets/static/Inter-Medium.ttf", import.meta.url))

export const badge = async (url: string) => {
	// initialize a new file system on each request to prevent cross request pollution
	const fs = Volume.fromJSON({})
	await clone(url, fs)
	// Set up the environment functions
	const env: EnvironmentFunctions = {
		$import: initialize$import({
			fs: patchedFs(fs.promises),
			fetch,
		}),
		$fs: patchedFs(fs.promises),
	}
	if (fs.existsSync("/inlang.config.js") === false) {
		// @felixhaeberle you should render a badge here that says "no inlang.config.js file found in the repository"
		throw new Error("No inlang.config.js file found in the repository.")
	}
	// Get the content of the inlang.config.js file
	const file = await fs.promises.readFile("/inlang.config.js", "utf-8")
	const { defineConfig } = await import(
		"data:application/javascript;base64," + btoa(file.toString())
	)
	const config: Config = await defineConfig(env)
	const resources = await config.readResources({ config })
	const [resourcesWithLints, errors] = await lint({ resources, config })
	if (errors) {
		console.error("lints partially failed", errors)
	}
	const lints = getLintReports(resourcesWithLints)
	console.log(lints)

	// @ts-ignore
	// const image = await satori(markup, {
	// 	width: 300,
	// 	height: 300,
	// 	fonts: [
	// 		{
	// 			family: "Inter",
	// 			weight: 400,
	// 			data: font,
	// 		},
	// 	],
	// })

	// return image
	return true
}

/**
 * Patching relative paths to absolute paths.
 *
 * Memfs does not support relative paths, so we need to patch them.
 */
const patchedFs = (fs: (typeof memfs)["promises"]) =>
	new Proxy(fs, {
		get: (target, prop) => {
			if (prop === "readFile") {
				return (path: string) => {
					if (path.startsWith("./")) {
						return fs.readFile(path.slice(1))
					}
					return fs.readFile(path)
				}
			} else if (prop === "readdir") {
				return (path: string, args: any) => {
					if (path.startsWith("./")) {
						return fs.readdir(path.slice(1), args)
					}
					return fs.readdir(path, args)
				}
			}
			return target[prop as keyof typeof target]
		},
	})
