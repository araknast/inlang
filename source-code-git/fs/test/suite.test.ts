import { test, expect } from "vitest"
import { memoryFs } from "src/memory"


// Basic test, should be split into several more elaborate tests later

test("memoryFs", async () => {
	const fs: Filesystem = new memoryFs()
	expect(await fs.readdir("/")).toEqual([])

	await fs.mkdir("home/user1/documents/")
	await fs.mkdir("home/user1/downloads")
	expect(await fs.readdir("/")).toEqual(["home"])
	expect(await fs.readdir("/home/user1/")).toEqual(["documents", "downloads"])

	await fs.writeFile("/home/user1/documents/file1", "text in the first file")
	await fs.writeFile("/file2", "text in the second file")
	expect(await fs.readdir("/home/user1/documents/")).toEqual(["file1"]);
	expect(await fs.readdir("/")).toEqual(["home", "file2"]);

	expect(await fs.readFile("/home/user1/documents/file1")).toEqual("text in the first file")
	expect(await fs.readFile("/file2")).toEqual("text in the second file")

	// relative paths
	await fs.writeFile("/home/user1/documents/../file3", "text in the third file")
	expect(await fs.readFile("./home/./user1/file3")).toEqual("text in the third file")

        const fsJson: string = await fs.toJson()
        // dirToJson
        expect(fsJson).toMatchSnapshot()
        
        // dirFromJson (circular references make this tricky to test)
        expect(
            JSON.stringify(await memoryFs.fromJson(JSON.stringify(fsJson)))
        ).toEqual(JSON.stringify(fs))

        expect (await fs.readFile("./home/dne")).toBeNull()
        expect (await fs.readdir("./home/dne")).toBeNull()
})
