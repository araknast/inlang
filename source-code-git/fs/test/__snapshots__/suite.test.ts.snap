// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`memoryFs 1`] = `
{
  "file2": "text in the second file",
  "home": {
    "user1": {
      "documents": {
        "file1": "text in the first file",
      },
      "downloads": {},
      "file3": "text in the third file",
    },
  },
}
`;
